src/ast/mod.rs

```diff
@@ -1101,405 +978,340 @@ pub enum Action {
-        match self {
-            Action::Let(lhs, rhs) => Action::Let(*lhs, f(rhs)),
-            Action::Set(lhs, args, rhs) => {
-                let right = f(rhs);
-                Action::Set(*lhs, args.iter().map(f).collect(), right)
-            }
-            Action::Delete(lhs, args) => Action::Delete(*lhs, args.iter().map(f).collect()),
-            Action::Union(lhs, rhs) => Action::Union(f(lhs), f(rhs)),
-            Action::Extract(expr, variants) => Action::Extract(f(expr), f(variants)),
-            Action::Panic(msg) => Action::Panic(msg.clone()),
-            Action::Expr(e) => Action::Expr(f(e)),
-        }
+    pub(crate) fn iter(&self) -> impl Iterator<Item = &GenericAction<Head, Leaf, Ann>> {
+        self.0.iter()
     }
 
-    pub fn replace_canon(&self, canon: &HashMap<Symbol, Expr>) -> Self {
-        match self {
-            Action::Let(lhs, rhs) => Action::Let(*lhs, rhs.subst(canon)),
-            Action::Set(lhs, args, rhs) => Action::Set(
-                *lhs,
-                args.iter().map(|e| e.subst(canon)).collect(),
-                rhs.subst(canon),
-            ),
-            Action::Delete(lhs, args) => {
-                Action::Delete(*lhs, args.iter().map(|e| e.subst(canon)).collect())
-            }
-            Action::Union(lhs, rhs) => Action::Union(lhs.subst(canon), rhs.subst(canon)),
-            Action::Extract(expr, variants) => {
-                Action::Extract(expr.subst(canon), variants.subst(canon))
-            }
-            Action::Panic(msg) => Action::Panic(msg.clone()),
-            Action::Expr(e) => Action::Expr(e.subst(canon)),
-        }
-    }
-}
-
-impl Display for NormAction {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "{}", self.to_action())
+    pub(crate) fn is_empty(&self) -> bool {
+        self.0.is_empty()
     }
 }
 
-impl Display for Action {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "{}", self.to_sexp())
-    }
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub enum CoreAction<Head, Leaf> {
+    Let(Leaf, Head, Vec<GenericAtomTerm<Leaf>>),
+    LetAtomTerm(Leaf, GenericAtomTerm<Leaf>),
+    Extract(GenericAtomTerm<Leaf>, GenericAtomTerm<Leaf>),
+    Set(Head, Vec<GenericAtomTerm<Leaf>>, GenericAtomTerm<Leaf>),
+    Delete(Head, Vec<GenericAtomTerm<Leaf>>),
+    Union(GenericAtomTerm<Leaf>, GenericAtomTerm<Leaf>),
+    Panic(String),
 }
 
-#[derive(Clone, Debug)]
-pub struct Rule {
-    // pub query: Query,
-    // pub actions: Vec<Action>,
-    pub head: Vec<Action>,
-    pub body: Vec<Fact>,
-}
+pub type NormAction = CoreAction<Symbol, Symbol>;
 
 #[derive(Clone, Debug, PartialEq, Eq, Hash)]
-pub struct NormRule {
-    pub head: Vec<NormAction>,
-    pub body: Vec<NormFact>,
-}
+pub struct CoreActions<Head, Leaf>(pub(crate) Vec<CoreAction<Head, Leaf>>);
+pub(crate) type NormActions = CoreActions<Symbol, Symbol>;
```

Comment:
do we use this anywhere? we could delete it

Reply:
It's used in `typechecking.rs`. Do you think if it's a good idea to change this to the `GenericCoreActions/CoreActions/ResolvedCoreActions` schema?

Reply:
Yeah that would make more sense